lib/container.js

```diff
@@ -164,8 +164,15 @@ export default class Container {
 
   logout() {
     let container = this;
-    this.clearCache();
-    return container.makeRequest('auth:logout', {}).then(function () {
+    let unregisterDevicePromise = Promise.resolve();
+    if (this.deviceID) {
+      unregisterDevicePromise = this.unregisterDevice();
+    }
+    return unregisterDevicePromise.then(function () {
+      container.clearCache();
+      return container.makeRequest('auth:logout', {});
+    })
```

Comment:
I am more favour in following logic to make the handling change as part of fails handle. Not prior check. And this part of code is actually duplicate of `if (!this.deviceID) {` check in  `unregisterDevice`

```
unregisterDevicePromise.then(function () {
      container.clearCache();
      return container.makeRequest('auth:logout', {});
}, function() {
// Check if the Error is missing device id?
// If yes, logout. else throw it.
}).then(function () {...});
````